### ISSUE TITLE
MemoryError: memory allocation failed error

### ISSUE THREAD

**Issue by KejaPower** [peterhinch/micropython-lcd160cr-gui#1]
Hello,
I'm using a pyboard v1.1 and trying to use the GUI to use a driver for a lora chip I get the following error : 
`Traceback (most recent call last):
  File "main.py", line 3, in <module>
  File "lcd_local.py", line 14, in <module>
MemoryError: memory allocation failed, allocating 155 bytes`

My main.py looks like this : 

`import gc
import ranger #driver for the lora chip
from lcd_local import setup
import font10
from gui.core.constants import *
from gui.core.lcd160_gui import Screen
from gui.widgets.buttons import Button
from gui.widgets.label import Label
from machine import SPI,Pin
from time import sleep
import micropython
from pyb import ExtInt
gc.collect()

micropython.alloc_emergency_exception_buf(100)
#initiate SPI

device_config = {
    'miso':"B14",
    'mosi':"B15",
    'ss':"B12",
    'sck':"B13",
    'busy':"B8",
    'dio_0':"C7"
}
reset=pyb.Pin.board.Y3

pins=device_config
device_spi = SPI(baudrate = 3500000,
        polarity = 0, phase = 0, bits = 8, firstbit = SPI.MSB,
        sck = Pin(device_config['sck'], Pin.OUT, Pin.PULL_DOWN),
        mosi = Pin(device_config['mosi'], Pin.OUT, Pin.PULL_UP),
        miso = Pin(device_config['miso'], Pin.IN, Pin.PULL_UP))

test=range.Ranger(device_spi,pins)

####GUI#########
def backbutton(x, y):
    def back(button):
        Screen.back()
    Button((x, y), font = font10, fontcolor = BLACK, callback = back,
           fgcolor = CYAN,  text = 'Back')

def setupbutton(x, y, cls_screen, text='Setup'):
    def setupbutton(button):
        Screen.change(cls_screen)
    Button((x, y), font = font10, callback = setupbutton, fgcolor = RED, text = text)

def quitbutton():
    def quit(button):
        Screen.shutdown()
    Button((109, 107), font = font10, callback = quit, fgcolor = RED, text = 'Quit')

def startbutton(x,y,text='Start'):
    def start(button):
        print('start')
    Button((x,y),font = font10, callback = start, fgcolor = BLUE, text = text)

class Screen_1(Screen):
    def __init__(self):
        super().__init__()
        Label((0, 0), font = font10, value = 'Configuration Menu')
        backbutton(0, 100)

class Screen_0(Screen):
    def __init__(self):
        super().__init__()
        Label((0, 0), font = font10, value = 'Main Menu')
        setupbutton(0, 107, Screen_1)
        startbutton(0,50)
        quitbutton()

setup()
Screen.change(Screen_0)`

I've tried to use gc.collect between the imports without luck. 

I must be doing something wrong with memorry allocation but I was not able to find how to fix it. The driver and the GUI work fine on their own.
Thanks for reading, any help would be appreciated :)

**Comment by peterhinch** [comment]
It seems that `import ranger` is using a lot of RAM, or causing RAM fragmentation, to the point where `lcd_local` can't even import the official lcd160cr driver. A first step is to read [this doc](http://docs.micropython.org/en/latest/reference/constrained.html).

One option is to issue `gc.collect()` after the import, but I don't think this will get you far enough. I would use frozen bytecode, freezing fonts, the GUI and the LORA libraries. Keep only your own code as normal Python modules to aid modification and debugging.

**Comment by KejaPower** [comment]
I had quite a lot of const in the ranger.py and I added a '_' for all of them.
I also changed the order of import in the main file: 
`import gc
import range
gc.collect()
import font10
from gui.core.constants import *
from gui.core.lcd160_gui import Screen
from gui.widgets.buttons import Button
from gui.widgets.label import Label
from machine import SPI,Pin
from time import sleep
import micropython
from pyb import ExtInt
from lcd_local import setup
gc.collect()`

and now it works fine.

I will still try to freeze fonts GUI and LORA librairies to bytecode to be safe.
thanks !
